### ISSUE TITLE
3D rendering seems to overwrite data array

### ISSUE THREAD

**Issue by li-li-github** [napari/napari#575]
## To Reproduce

1. Display a 3D data on napari.
2. Adjust contrast.
3. Contrast does not come back even when you tune the contrast bar back to original position.

## Environment

 - napari Version (e.g., 1.0): 0.2.0
 - OS (e.g., Linux): Mac OS 10.14.6
 - Python version: 3.6.7
 - Any other relevant information:

**Comment by jni** [comment]
@li-li-github I can't reproduce this on my machine. At any rate I think and hope that the error might be in the display, not in modifying the data.

Could you provide a complete example with data? In the code below, I can adjust the contrast back and forth perfectly in the 3D display:

```python
import numpy as np
from skimage import data
from skimage.util import random_noise
import napari

blobs_raw = np.stack([data.binary_blobs(length=64, n_dim=3, volume_fraction=f)
                      for f in np.linspace(0.05, 0.5, 10)])
blobs_sp = random_noise(blobs_raw, mode='s&p')
blobs_g = random_noise(blobs_sp, mode='gaussian')
blobs = random_noise(blobs_g, mode='poisson')
with napari.gui_qt():
    viewer = napari.view_image(blobs)
    viewer.dims.ndisplay = 3
```

**Comment by li-li-github** [comment]
Hi @jni , thank you for the prompt reply.
I found that it doesn't work properly when datatype is `np.float32`.
I have tested this with other np.floats and found only float32 doesn't work properly.

**Comment by sofroniewn** [comment]
@jni @li-li-github I am able to reproduce this bug by modifying the code from @jni to be
`viewer = napari.view_image(blobs.astype(np.float32))`
and then moving the left clim handle all the way to the right - my screen becomes black, and the data will never reappear. It appears to be a divide by 0 error within vispy that vispy can never recover from. Here is the error message I see
```python
/Users/nicholassofroniew/Github/vispy/vispy/visuals/volume.py:483: RuntimeWarning: overflow encountered in true_divide
  vol /= self._clim[1] - self._clim[0]
```

Ultimately this will be fixed by a PR to vispy which we are happy to make, in the mean time we can look into some workarounds / fixes on our end, and I recommend trying to avoid shrinking the clim range to be too small.

Thanks for catching this @li-li-github!!

**Comment by jni** [comment]
Just double checked, and I can reproduce @li-li-github's original issues (overwriting data), and this has nothing to do with clims (at least that I can tell): the input array is indeed overwritten. To reproduce the issue:

1. Download @VolkerH's data: https://www.dropbox.com/s/93ykwmonofq7171/droso_ovarioles.tif?dl=0
2. Load it with `skimage.io.imread`. Check the range: it is a float32 array with range [0, 255.0].
3. Call napari.view_image on it
4. Set viewer.ndims.ndisplay = 3
5. Check the range of the original image. It is now in [0, (something times 10^-11)]. 

I'd consider this a critical bug.

**Comment by tlambert03** [comment]
figured it out.  commenting here rather than #612 to keep the discussion in one place, but refer there for test code if you want.

The problem is [these lines](https://github.com/vispy/vispy/blob/master/vispy/visuals/volume.py#L477-L483) in the vispy `VolumeVisual.set_data()` method:
```python
vol = np.array(vol, dtype='float32', copy=False)
if self._clim[1] == self._clim[0]:
    if self._clim[0] != 0.:
        vol *= 1.0 / self._clim[0]
else:
    vol -= self._clim[0]
    vol /= self._clim[1] - self._clim[0]
```
with all datatypes _besides_ `float32`, it has no choice but to make a copy of the array during type conversion, so the memory address of the `vol` variable changes.  However, with `float32` and `copy=False`, the memory address stays the same, and when the `clim` gets applied, it actually changes the data that was delivered in the `VispyImageLayer._on_data_change` method. This only happens with 3D `VolumeVisuals`, since the set_data method for ImageVisuals don't normalize the data in the same way. 

Additionally, there are multiple events that are directly or indirectly connected to the `VispyImageLayer._on_data_change` method, so that method actually gets called **6 times** for each layer when you click the `ndisplay` button once... and each time the volume gets renormalized to the `clim` driving it down to those crazy low numbers.  (and probably incurring some additional overhead outside of this problem)

So [that line ](https://github.com/vispy/vispy/blob/master/vispy/visuals/volume.py#L477) in vispy will either need to change, or we will need to just force it to copy `float32` arrays before calling `node.set_data()` as I did in my fixed branch:
https://github.com/napari/napari/blob/efb55e23599a412da2a5f3ed1a807823c42fe96b/napari/_vispy/vispy_image_layer.py#L70-L76

if you agree, let me know and I'll send a PR including the test and a fix.

**Comment by jni** [comment]
@tlambert03 nice sleuthing! We should definitely make a float64 copy of the array on our end temporarily (a float32 copy will still get downscaled repeatedly and give a black image, even if our original data is at least not destroyed, right?), *and* VisPy should be fixed so that this doesn't happen again. But we certainly can't wait for the VisPy fix to be out in the wild.

btw the 6 times calling of on_data_change definitely needs to be fixed also...! 😬

**Comment by tlambert03** [comment]
The code I posted above doesn't repeatedly downscale (as long as that `data.copy()` is in there for the `dtype == 'float32'` case) since it gets "reset" each time the `VispyImageLayer._on_data_change` method gets called.    Whereas, storing a float64 would increase the memory overhead, and still wouldn't avoid the `copy` event that would then just happen in the vispy code anyway.

I'm not sure vispy would necessarily see it as a "bug" (one can imagine a case in which it's better not to duplicate the data when setting it...).  They might consider exposing that `copy=False` parameter, but it's also easy enough just to copy the data ourselves before calling `set_data` and the effect should be the same.

**Comment by jni** [comment]
Ah, gotcha. Well, either way, making six copies of the data every time something changes seems like a bad idea. =P

I definitely think it's a bug on the vispy side: visualisation code should *never* modify input data. I think I'm quite happy to die on this hill. =P

**Comment by tlambert03** [comment]
I guess I'll join you on that hill.  see vispy/vispy#1727 if you want to chime in

**Comment by sofroniewn** [comment]
Apologies for the six calls to `on_data_change`, yikes, I knew there might be some stuff like that happening in the repo, but that exceeds my fears!! We've recently started adding some benchmarks - see our [benchmarking guide](https://github.com/napari/napari/blob/master/docs/BENCHMARKS.md) which can also be profiled, and I feel like now is a good time to start reducing some of those redundant calls, but that will be separate chunks of work that we can discuss elsewhere

I'm going to close this issue now that we've merged #613 and started a discussion on the vispy repo too - thanks @tlambert03

**Comment by tlambert03** [comment]
This is now fixed upstream in vispy/vispy@876a0d3 ...  so once v0.6.2 is released, in order to avoid the extra copy operation for float32, we'll either need to pin a higher vispy dependency and remove the conditional added in #613, or dynamically check vispy versions inline.  I'll try to remember to open a new issue when that happens

**Comment by jni** [comment]
Since we are still in alpha we are in general totally fine with being bleeding edge with dependencies. We'll bump vispy once it comes out. Thanks for your lightning quick handling of this, @tlambert03!
